src/CycloneDX.Core/Json/Validator.cs

```diff
@@ -190,6 +284,33 @@ private static ValidationResult Validate(JsonSchema schema, JsonDocument jsonDoc
                         }
                     }
                 }
+
+                // The JSON Schema, at least the ones defined by CycloneDX
+                // and handled by current parser in dotnet ecosystem, can
+                // not specify or check the uniqueness requirement for the
+                // "bom-ref" assignments in the overall document (e.g. in
+                // "metadata/component" and list of "components", as well
+                // as in "services" and "vulnerabilities", as of CycloneDX
+                // spec v1.4), so this is checked separately here if the
+                // document seems structurally intact otherwise.
+                // Note that this is not a problem for the XML schema with
+                // its explicit <xs:unique name="bom-ref"> constraint.
+                Dictionary<string, List<JsonElement>> bomRefs = findNamedElements(jsonDocument.RootElement, "bom-ref");
+                foreach (KeyValuePair<string, List<JsonElement>> KVP in bomRefs) {
+                    if (KVP.Value != null && KVP.Value.Count != 1) {
+                        validationMessages.Add($"'bom-ref' value of {KVP.Key}: expected 1 mention, actual {KVP.Value.Count}");
+                    }
+                }
+
+                // Check that if we "ref" something (from dependencies, etc.)
+                // the corresponding "bom-ref" exists in this document:
+                List<string> bomRefsList = new List<string>(bomRefs.Keys);
+                Dictionary<string, List<JsonElement>> useRefs = findNamedElements(jsonDocument.RootElement, "ref");
+                foreach (KeyValuePair<string, List<JsonElement>> KVP in useRefs) {
+                    if (KVP.Value != null && KVP.Value.Count > 0 && !(bomRefsList.Contains(KVP.Key))) {
+                        validationMessages.Add($"'ref' value of {KVP.Key} was used in {KVP.Value.Count} place(s); expected a 'bom-ref' defined for it, but there was none");
+                    }
+                }
```

Comment:
This needs to support bom-link references, that won't be defined in the current document.

https://cyclonedx.org/capabilities/bomlink/

Reply:
Thanks, a nice facility that I did not realize. Would it suffice to just ignore here `ref` values by simple `.StartsWith()`? And of what prefix (`urn:` seems generic enough for any schema expansions...)? Or should it match specifically that standard syntax (including some regex of UUID) - IMHO a slower, cumbersome to code, and less future-proof approach?

Reply:
Correct me if I'm wrong but it depends on the location of "ref" in the SBOM whether a bomlink is allowed or not.
https://github.com/CycloneDX/specification/blob/master/schema/bom-1.5.schema.json
In "dependency" only refLinkType is allowed, in other places like:
- vulnerability/affects
- components/modelCard/modelParameters/datasets
- formulation/workflows/resourceReferences

both refLinkType as well as bomLinkElementType are allowed.
(As a matter of fact there are several references which are not called "ref", like
- composition/assemblies
- components/evidence/identity/tools
- annotations/subjects

)

Shouldn't thus there be a requirement for "bom-ref" that it shouldn't have the form of a bomlink? This is more a question about the specification. Otherwise, you can't distinguish the two. (Of course, this might be a rather theoretical issue.)

Reply:
Hm, I'm in favor of reverting this last commit and putting it into another PR for you guys to lawyer it out :D and not block this one.

I don't think BTW there were similar constraints for XML validation (XSD mostly)? So probably a holistic fix to validate this should take care of both areas (so more work than I can spare now).

Reply:
Offloaded into https://github.com/jimklimov/cyclonedx-dotnet-library/tree/validate-json-bomref-usedRef, and reverted this PR branch to one commit before it, so there is no immediate flaw blocking the merge :D